mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java

```diff
@@ -154,6 +155,44 @@ public MongoDBDialect(MongoDBDatastoreProvider provider) {
 	@Override
 	public Tuple getTuple(EntityKey key, TupleContext tupleContext) {
 		DBObject found = this.getObject( key, tupleContext );
+		return createTuple( key, tupleContext, found );
+	}
+
+	@Override
+	public List<Tuple> getTuples(EntityKey[] keys, TupleContext tupleContext) {
+		DBCursor cursor = this.getObjects( keys, tupleContext );
+		List<Tuple> tuples = createListOfTuples( keys );
+		try {
+			for ( DBObject dbObject : cursor ) {
+				for ( int i = 0; i < keys.length; i++ ) {
+					if ( matches( dbObject, keys[i].getColumnNames(), keys[i].getColumnValues() ) ) {
+						tuples.set( i, createTuple( keys[i], tupleContext, dbObject ) );
+					}
+				}
+			}
+			return tuples;
+		}
+		finally {
+			if ( cursor != null ) {
+				cursor.close();
+			}
+		}
+	}
+
+	private List<Tuple> createListOfTuples(EntityKey[] keys) {
+		List<Tuple> tuples = new ArrayList<>( keys.length );
+		for ( int i = 0; i < keys.length; i++ ) {
```

Comment:
Thinking more about it, the two nested loops seem not optimal. If you could ensure the order of results matches the order of keys (e.g. by applying a sort by id), it could be avoided by processing the results like so:

    while ( result.hasNext() ) {
        advance incoming key iterator to matching key
        add null to result list for each advancing without matching
        when find the matching key, add the result
        advance result iterator
    }

That way you'd iterate the keys once and the results once and not the keys once per result.